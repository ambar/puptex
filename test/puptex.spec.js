const puptex = require('../src/puptex')

describe('puptex', () => {
  beforeAll(async () => {
    puptex.config({
      concurrency: 1,
      mathJax: {
        TeX: {
          Macros: {
            bm: '\\boldsymbol',
          },
        },
      },
    })
    await puptex.launch()
  })

  afterAll(() => {
    puptex.close()
  })

  it('renders empty', async () => {
    const data = await puptex.renderMath('')
    expect(data).toMatchSnapshot()
  })

  it('renders tex', async () => {
    const data = await puptex.renderMath('E=mc^2')
    expect(data).toMatchSnapshot()
  })

  it('renders `parse error`', async () => {
    const data = await puptex.renderMath('E=mc^')
    expect(data).toMatchSnapshot()
  })

  it('clears global glyphs cache', async () => {
    const math = 'E=mc^2'
    const [a, b] = await Promise.all([
      puptex.renderMath(math),
      puptex.renderMath(math),
    ])
    expect(a.svg).toEqual(b.svg)
  })

  it('renders with `16px` font-size', async () => {
    const data = await puptex.renderMath('', {fontSize: '16px'})
    expect(data).toMatchSnapshot()
  })

  it('renders CJK characters correctly', async () => {
    const data = await puptex.renderMath('\\text{拾玖}^i')
    expect(data).toMatchSnapshot()
  })

  it('renders `\\bm`', async () => {
    const data = await puptex.renderMath('\\bm{x}')
    expect(data).toMatchSnapshot()
  })

  // it('renders MathML', async () => {
  //   const data = await puptex.renderMath('x', {mml: true})
  //   expect(data).toMatchSnapshot()
  // })
})
