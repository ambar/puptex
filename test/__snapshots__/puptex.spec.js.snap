// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`puptex renders \`\\bm\` 1`] = `
Object {
  "height": "1.548ex",
  "style": "vertical-align: -0.284ex; font-size: 15px;",
  "svg": "<svg xmlns:xlink=\\"http://www.w3.org/1999/xlink\\" width=\\"1.532ex\\" height=\\"1.548ex\\" viewBox=\\"0 -544 659.5 666.4\\" role=\\"img\\" focusable=\\"false\\" xmlns=\\"http://www.w3.org/2000/svg\\" style=\\"vertical-align: -0.284ex; font-size: 15px;\\"><defs><path stroke-width=\\"1\\" id=\\"E1-MJMATHBI-78\\" d=\\"M74 282H63Q43 282 43 296Q43 298 45 307T56 332T76 365T110 401T159 433Q200 451 233 451H236Q273 451 282 450Q358 437 382 400L392 410Q434 452 483 452Q538 452 568 421T599 346Q599 303 573 280T517 256Q494 256 478 270T462 308Q462 343 488 367Q501 377 520 385Q520 386 516 389T502 396T480 400T462 398Q429 383 415 341Q354 116 354 80T405 44Q449 44 485 74T535 142Q539 156 542 159T562 162H568H579Q599 162 599 148Q599 135 586 111T550 60T485 12T397 -8Q313 -8 266 35L258 44Q215 -7 161 -7H156Q99 -7 71 25T43 95Q43 143 70 165T125 188Q148 188 164 174T180 136Q180 101 154 77Q141 67 122 59Q124 54 136 49T161 43Q183 43 200 61T226 103Q287 328 287 364T236 400Q200 400 164 377T107 302Q103 288 100 285T80 282H74Z\\"></path></defs><g stroke=\\"currentColor\\" fill=\\"currentColor\\" stroke-width=\\"0\\" transform=\\"matrix(1 0 0 -1 0 0)\\"><use xmlns:xlink=\\"http://www.w3.org/1999/xlink\\" xlink:href=\\"#E1-MJMATHBI-78\\" x=\\"0\\" y=\\"0\\"></use></g></svg>",
  "width": "1.532ex",
}
`;

exports[`puptex renders \`parse error\` 1`] = `
Object {
  "errors": Array [
    "TeX Jax - parse error: Missing superscript or subscript argument",
  ],
  "height": "3.786ex",
  "style": "vertical-align: -1.263ex; margin-left: -0.012ex; font-size: 15px;",
  "svg": "<svg xmlns:xlink=\\"http://www.w3.org/1999/xlink\\" width=\\"35.006ex\\" height=\\"3.786ex\\" viewBox=\\"-5 -1085.9 15072.1 1629.9\\" role=\\"img\\" focusable=\\"false\\" xmlns=\\"http://www.w3.org/2000/svg\\" style=\\"vertical-align: -1.263ex; margin-left: -0.012ex; font-size: 15px;\\"><defs></defs><g stroke=\\"currentColor\\" fill=\\"currentColor\\" stroke-width=\\"0\\" transform=\\"matrix(1 0 0 -1 0 0)\\"><rect fill=\\"rgb(255, 255, 136)\\" stroke=\\"none\\" width=\\"15062\\" height=\\"1427\\" x=\\"0\\" y=\\"-429\\"></rect><g fill=\\"rgb(204, 0, 0)\\" stroke=\\"rgb(204, 0, 0)\\" transform=\\"translate(240,0)\\"><g transform=\\"scale(0.9)\\"><text direction=\\"ltr\\" stroke=\\"none\\" transform=\\"scale(60.217) matrix(1 0 0 -1 0 0)\\">Missing superscript or subscript argument</text></g></g><line stroke-linecap=\\"square\\" stroke=\\"rgb(204, 0, 0)\\" stroke-width=\\"60.22\\" y1=\\"30\\" x2=\\"30\\" x1=\\"30\\" y2=\\"1397\\" transform=\\"translate(-5,-429)\\"></line><line stroke-linecap=\\"square\\" stroke=\\"rgb(204, 0, 0)\\" stroke-width=\\"60.22\\" y1=\\"30\\" x2=\\"30\\" x1=\\"30\\" y2=\\"1397\\" transform=\\"translate(15006,-429)\\"></line><line stroke-linecap=\\"square\\" stroke=\\"rgb(204, 0, 0)\\" stroke-width=\\"60.22\\" y2=\\"30\\" y1=\\"30\\" x1=\\"30\\" x2=\\"15031\\" transform=\\"translate(0,944)\\"></line><line stroke-linecap=\\"square\\" stroke=\\"rgb(204, 0, 0)\\" stroke-width=\\"60.22\\" y2=\\"30\\" y1=\\"30\\" x1=\\"30\\" x2=\\"15031\\" transform=\\"translate(0,-434)\\"></line></g></svg>",
  "width": "35.006ex",
}
`;

exports[`puptex renders CJK characters correctly 1`] = `
Object {
  "height": "4.345ex",
  "style": "vertical-align: -1.263ex; font-size: 15px;",
  "svg": "<svg xmlns:xlink=\\"http://www.w3.org/1999/xlink\\" width=\\"5.275ex\\" height=\\"4.345ex\\" viewBox=\\"0 -1326.8 2271.3 1870.7\\" role=\\"img\\" focusable=\\"false\\" xmlns=\\"http://www.w3.org/2000/svg\\" style=\\"vertical-align: -1.263ex; font-size: 15px;\\"><defs><path stroke-width=\\"1\\" id=\\"E1-MJMATHI-69\\" d=\\"M184 600Q184 624 203 642T247 661Q265 661 277 649T290 619Q290 596 270 577T226 557Q211 557 198 567T184 600ZM21 287Q21 295 30 318T54 369T98 420T158 442Q197 442 223 419T250 357Q250 340 236 301T196 196T154 83Q149 61 149 51Q149 26 166 26Q175 26 185 29T208 43T235 78T260 137Q263 149 265 151T282 153Q302 153 302 143Q302 135 293 112T268 61T223 11T161 -11Q129 -11 102 10T74 74Q74 91 79 106T122 220Q160 321 166 341T173 380Q173 404 156 404H154Q124 404 99 371T61 287Q60 286 59 284T58 281T56 279T53 278T49 278T41 278H27Q21 284 21 287Z\\"></path></defs><g stroke=\\"currentColor\\" fill=\\"currentColor\\" stroke-width=\\"0\\" transform=\\"matrix(1 0 0 -1 0 0)\\"><text font-family=\\"STIXGeneral,'Arial Unicode MS',serif\\" stroke=\\"none\\" transform=\\"scale(60.217) matrix(1 0 0 -1 0 0)\\">拾</text><g transform=\\"translate(963,0)\\"><text font-family=\\"STIXGeneral,'Arial Unicode MS',serif\\" stroke=\\"none\\" transform=\\"scale(60.217) matrix(1 0 0 -1 0 0)\\">玖</text></g><use transform=\\"scale(0.707)\\" xmlns:xlink=\\"http://www.w3.org/1999/xlink\\" xlink:href=\\"#E1-MJMATHI-69\\" x=\\"2725\\" y=\\"1061\\"></use></g></svg>",
  "width": "5.275ex",
}
`;

exports[`puptex renders MathML 1`] = `
Object {
  "height": "1.548ex",
  "mml": "<math xmlns=\\"http://www.w3.org/1998/Math/MathML\\" display=\\"block\\"><mi>x</mi></math>",
  "style": "vertical-align: -0.284ex; font-size: 15px;",
  "svg": "<svg xmlns:xlink=\\"http://www.w3.org/1999/xlink\\" width=\\"1.33ex\\" height=\\"1.548ex\\" viewBox=\\"0 -544 572.5 666.4\\" role=\\"img\\" focusable=\\"false\\" xmlns=\\"http://www.w3.org/2000/svg\\" style=\\"vertical-align: -0.284ex; font-size: 15px;\\"><defs><path stroke-width=\\"1\\" id=\\"E1-MJMATHI-78\\" d=\\"M52 289Q59 331 106 386T222 442Q257 442 286 424T329 379Q371 442 430 442Q467 442 494 420T522 361Q522 332 508 314T481 292T458 288Q439 288 427 299T415 328Q415 374 465 391Q454 404 425 404Q412 404 406 402Q368 386 350 336Q290 115 290 78Q290 50 306 38T341 26Q378 26 414 59T463 140Q466 150 469 151T485 153H489Q504 153 504 145Q504 144 502 134Q486 77 440 33T333 -11Q263 -11 227 52Q186 -10 133 -10H127Q78 -10 57 16T35 71Q35 103 54 123T99 143Q142 143 142 101Q142 81 130 66T107 46T94 41L91 40Q91 39 97 36T113 29T132 26Q168 26 194 71Q203 87 217 139T245 247T261 313Q266 340 266 352Q266 380 251 392T217 404Q177 404 142 372T93 290Q91 281 88 280T72 278H58Q52 284 52 289Z\\"></path></defs><g stroke=\\"currentColor\\" fill=\\"currentColor\\" stroke-width=\\"0\\" transform=\\"matrix(1 0 0 -1 0 0)\\"><use xmlns:xlink=\\"http://www.w3.org/1999/xlink\\" xlink:href=\\"#E1-MJMATHI-78\\" x=\\"0\\" y=\\"0\\"></use></g></svg>",
  "width": "1.33ex",
}
`;

exports[`puptex renders empty 1`] = `
Object {
  "height": "0.289ex",
  "style": "vertical-align: -0.145ex; font-size: 15px;",
  "svg": "<svg xmlns:xlink=\\"http://www.w3.org/1999/xlink\\" width=\\"0\\" height=\\"0.289ex\\" viewBox=\\"0 -62.2 0 124.4\\" role=\\"img\\" focusable=\\"false\\" xmlns=\\"http://www.w3.org/2000/svg\\" style=\\"vertical-align: -0.145ex; font-size: 15px;\\"><defs></defs><g stroke=\\"currentColor\\" fill=\\"currentColor\\" stroke-width=\\"0\\" transform=\\"matrix(1 0 0 -1 0 0)\\"></g></svg>",
  "width": "0",
}
`;

exports[`puptex renders tex 1`] = `
Object {
  "height": "2.527ex",
  "style": "vertical-align: -0.284ex; font-size: 15px;",
  "svg": "<svg xmlns:xlink=\\"http://www.w3.org/1999/xlink\\" width=\\"8.976ex\\" height=\\"2.527ex\\" viewBox=\\"0 -965.5 3864.5 1087.9\\" role=\\"img\\" focusable=\\"false\\" xmlns=\\"http://www.w3.org/2000/svg\\" style=\\"vertical-align: -0.284ex; font-size: 15px;\\"><defs><path stroke-width=\\"1\\" id=\\"E1-MJMATHI-45\\" d=\\"M492 213Q472 213 472 226Q472 230 477 250T482 285Q482 316 461 323T364 330H312Q311 328 277 192T243 52Q243 48 254 48T334 46Q428 46 458 48T518 61Q567 77 599 117T670 248Q680 270 683 272Q690 274 698 274Q718 274 718 261Q613 7 608 2Q605 0 322 0H133Q31 0 31 11Q31 13 34 25Q38 41 42 43T65 46Q92 46 125 49Q139 52 144 61Q146 66 215 342T285 622Q285 629 281 629Q273 632 228 634H197Q191 640 191 642T193 659Q197 676 203 680H757Q764 676 764 669Q764 664 751 557T737 447Q735 440 717 440H705Q698 445 698 453L701 476Q704 500 704 528Q704 558 697 578T678 609T643 625T596 632T532 634H485Q397 633 392 631Q388 629 386 622Q385 619 355 499T324 377Q347 376 372 376H398Q464 376 489 391T534 472Q538 488 540 490T557 493Q562 493 565 493T570 492T572 491T574 487T577 483L544 351Q511 218 508 216Q505 213 492 213Z\\"></path><path stroke-width=\\"1\\" id=\\"E1-MJMAIN-3D\\" d=\\"M56 347Q56 360 70 367H707Q722 359 722 347Q722 336 708 328L390 327H72Q56 332 56 347ZM56 153Q56 168 72 173H708Q722 163 722 153Q722 140 707 133H70Q56 140 56 153Z\\"></path><path stroke-width=\\"1\\" id=\\"E1-MJMATHI-6D\\" d=\\"M21 287Q22 293 24 303T36 341T56 388T88 425T132 442T175 435T205 417T221 395T229 376L231 369Q231 367 232 367L243 378Q303 442 384 442Q401 442 415 440T441 433T460 423T475 411T485 398T493 385T497 373T500 364T502 357L510 367Q573 442 659 442Q713 442 746 415T780 336Q780 285 742 178T704 50Q705 36 709 31T724 26Q752 26 776 56T815 138Q818 149 821 151T837 153Q857 153 857 145Q857 144 853 130Q845 101 831 73T785 17T716 -10Q669 -10 648 17T627 73Q627 92 663 193T700 345Q700 404 656 404H651Q565 404 506 303L499 291L466 157Q433 26 428 16Q415 -11 385 -11Q372 -11 364 -4T353 8T350 18Q350 29 384 161L420 307Q423 322 423 345Q423 404 379 404H374Q288 404 229 303L222 291L189 157Q156 26 151 16Q138 -11 108 -11Q95 -11 87 -5T76 7T74 17Q74 30 112 181Q151 335 151 342Q154 357 154 369Q154 405 129 405Q107 405 92 377T69 316T57 280Q55 278 41 278H27Q21 284 21 287Z\\"></path><path stroke-width=\\"1\\" id=\\"E1-MJMATHI-63\\" d=\\"M34 159Q34 268 120 355T306 442Q362 442 394 418T427 355Q427 326 408 306T360 285Q341 285 330 295T319 325T330 359T352 380T366 386H367Q367 388 361 392T340 400T306 404Q276 404 249 390Q228 381 206 359Q162 315 142 235T121 119Q121 73 147 50Q169 26 205 26H209Q321 26 394 111Q403 121 406 121Q410 121 419 112T429 98T420 83T391 55T346 25T282 0T202 -11Q127 -11 81 37T34 159Z\\"></path><path stroke-width=\\"1\\" id=\\"E1-MJMAIN-32\\" d=\\"M109 429Q82 429 66 447T50 491Q50 562 103 614T235 666Q326 666 387 610T449 465Q449 422 429 383T381 315T301 241Q265 210 201 149L142 93L218 92Q375 92 385 97Q392 99 409 186V189H449V186Q448 183 436 95T421 3V0H50V19V31Q50 38 56 46T86 81Q115 113 136 137Q145 147 170 174T204 211T233 244T261 278T284 308T305 340T320 369T333 401T340 431T343 464Q343 527 309 573T212 619Q179 619 154 602T119 569T109 550Q109 549 114 549Q132 549 151 535T170 489Q170 464 154 447T109 429Z\\"></path></defs><g stroke=\\"currentColor\\" fill=\\"currentColor\\" stroke-width=\\"0\\" transform=\\"matrix(1 0 0 -1 0 0)\\"><use xmlns:xlink=\\"http://www.w3.org/1999/xlink\\" xlink:href=\\"#E1-MJMATHI-45\\" x=\\"0\\" y=\\"0\\"></use><use xmlns:xlink=\\"http://www.w3.org/1999/xlink\\" xlink:href=\\"#E1-MJMAIN-3D\\" x=\\"1042\\" y=\\"0\\"></use><use xmlns:xlink=\\"http://www.w3.org/1999/xlink\\" xlink:href=\\"#E1-MJMATHI-6D\\" x=\\"2098\\" y=\\"0\\"></use><g transform=\\"translate(2977,0)\\"><use xmlns:xlink=\\"http://www.w3.org/1999/xlink\\" xlink:href=\\"#E1-MJMATHI-63\\" x=\\"0\\" y=\\"0\\"></use><use transform=\\"scale(0.707)\\" xmlns:xlink=\\"http://www.w3.org/1999/xlink\\" xlink:href=\\"#E1-MJMAIN-32\\" x=\\"613\\" y=\\"583\\"></use></g></g></svg>",
  "width": "8.976ex",
}
`;

exports[`puptex renders with \`16px\` font-size 1`] = `
Object {
  "height": "0.289ex",
  "style": "vertical-align: -0.145ex; font-size: 16px;",
  "svg": "<svg xmlns:xlink=\\"http://www.w3.org/1999/xlink\\" width=\\"0\\" height=\\"0.289ex\\" viewBox=\\"0 -62.2 0 124.4\\" role=\\"img\\" focusable=\\"false\\" xmlns=\\"http://www.w3.org/2000/svg\\" style=\\"vertical-align: -0.145ex; font-size: 16px;\\"><defs></defs><g stroke=\\"currentColor\\" fill=\\"currentColor\\" stroke-width=\\"0\\" transform=\\"matrix(1 0 0 -1 0 0)\\"></g></svg>",
  "width": "0",
}
`;
